### PR TITLE
[AArch64] Support optional constant offset for constraint "S"

### PR DESCRIPTION
**Pull request by MaskRay**
Modify the initial implementation (https://reviews.llvm.org/D46745) to
support a constant offset so that the following code will compile:
```
int a[2][2];
void foo() { asm("// %0" :: "S"(&a[1][1])); }
```

We use the generic code path for "s". In GCC's aarch64 port, "S" is
supported for PIC while "s" isn't, making "s" less useful. We implement
"S" but not "s".

Similar to #80201 for RISC-V.

### CODE REVIEW

**Comment by MaskRay** [comment]
I have some notes about these constraints https://maskray.me/blog/2024-01-30-raw-symbol-names-in-inline-assembly

**Comment by smithp35** [comment]
Just to make sure I've understood:
* x86_64's support for constraint "s" was hoisted into the target independent TargetLowering in https://reviews.llvm.org/D61560 
* Clang's implementation of constraint "s" isn't restricted to non-PIC/non-PIE like GCC's
* We can implement constraint "S" in terms of constraint "s" as Clang's implementation of "s" is equivalent to AArch64 "S".

On that basis this looks reasonable to me. I've not had time to pick over the details of the code/test changes though.

**Comment by MaskRay** [comment]
> Just to make sure I've understood:
> 
> * x86_64's support for constraint "s" was hoisted into the target independent TargetLowering in [reviews.llvm.org/D61560](https://reviews.llvm.org/D61560)
> * Clang's implementation of constraint "s" isn't restricted to non-PIC/non-PIE like GCC's
> * We can implement constraint "S" in terms of constraint "s" as Clang's implementation of "s" is equivalent to AArch64 "S".

Yes, these are all correct. 

> On that basis this looks reasonable to me. I've not had time to pick over the details of the code/test changes though.

Thanks!

File: llvm/test/CodeGen/RISCV/inline-asm-S-constraint.ll [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ; RUN: llc -mtriple=riscv64 < %s | FileCheck %s --check-prefix=RV64
 
 @var = external dso_local global i32, align 4
+@a = external global [2 x [2 x i32]], align 4
```
Comment (smithp35) [inline llvm/test/CodeGen/RISCV/inline-asm-S-constraint.ll:6]: Did you mean to change this file in this PR? I think this change has already been made in #80201

**Review by smithp35 — APPROVED** [review]
Thanks for the response. LGTM.

I think the RISCV inline-asm-S-constraint.ll is no longer necessary thanks to #80201

File: llvm/test/CodeGen/RISCV/inline-asm-S-constraint.ll (hunk 1 above)
Comment (MaskRay) [reply llvm/test/CodeGen/RISCV/inline-asm-S-constraint.ll:6]: Sorry, no. I accidentally included this when preparing tests using my experience on x86 Ws and RISC-V S:)

**Pull request merged**